From 4b90028daea354041ff3e9d63752019f30f69760 Mon Sep 17 00:00:00 2001
From: Oleksandr Shamray <oleksandrs@nvidia.com>
Date: Wed, 8 Nov 2023 14:52:38 +0200
Subject: [PATCH 64/78] hwmon: (mlxsw): Downstream: Allow fan speed setting
 granularity of 1 PWM

Currently PWM setting is allowed with 10 percent stepping.
Such configuration is aligned with thermal drivers, which are used to be
bound to "mlxsw" driver.

This binding happens when the cooling instances created by the driver are
bound to some kernel thermal driver.

In case system is not using kernel thermal control and the cooling
instances created by the driver are not bound to any thermal driver, the
driver still does not allow setting of PWM granularity less than 10
percent.

Allow setting fan with one percent granularity, thus any user space
thermal application will be able to set PWM to any allowed value in range
from 51 PWM to 255 PWM.

Note: this is downstream patch, since it can affect functionality for
the Nvidia users running kernel thermal control. So, it is not going to be
submitted to up-stream.

Signed-off-by: Oleksandr Shamray <oleksandrs@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/core_thermal.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c b/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
index e746cd9..e19f36f2e 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/core_thermal.c
@@ -24,8 +24,13 @@
 #define MLXSW_THERMAL_MODULE_TEMP_HOT	80000	/* 80C */
 #define MLXSW_THERMAL_HYSTERESIS_TEMP	5000	/* 5C */
 #define MLXSW_THERMAL_MODULE_TEMP_SHIFT	(MLXSW_THERMAL_HYSTERESIS_TEMP * 2)
+#ifdef CONFIG_MLXSW_CORE_THERMAL
 #define MLXSW_THERMAL_MAX_STATE	10
 #define MLXSW_THERMAL_MIN_STATE	2
+#else
+#define MLXSW_THERMAL_MAX_STATE	255
+#define MLXSW_THERMAL_MIN_STATE	51	/* 20 percent */
+#endif
 #define MLXSW_THERMAL_MAX_DUTY	255
 
 /* External cooling devices, allowed for binding to mlxsw thermal zones. */
-- 
2.8.4

